common/libs/wallet/non-deterministic/parity.ts

```diff
@@ -0,0 +1,59 @@
+import EthTx from 'ethereumjs-tx';
+import { translateRaw } from 'translations';
+import { IFullWallet } from '../IWallet';
+import { TSetWalletQrTransaction, TFinalizeQrTransaction } from 'actions/wallet';
+
+export default class ParityWallet implements IFullWallet {
+  public address: string;
+  private setWalletQrTransaction: TSetWalletQrTransaction;
+  private finalizeQrTransaction: TFinalizeQrTransaction;
+
+  constructor(
+    address: string,
+    setWalletQrTransaction: TSetWalletQrTransaction,
+    finalizeQrTransaction: TFinalizeQrTransaction
+  ) {
+    this.address = address;
+    this.setWalletQrTransaction = setWalletQrTransaction;
+    this.finalizeQrTransaction = finalizeQrTransaction;
+  }
+
+  public signRawTransaction(tx: EthTx): Promise<Buffer> {
+    return new Promise((resolve, reject) => {
+      const from = this.address;
+
+      const onCancel = () => {
+        this.finalizeQrTransaction();
+        reject(new Error(translateRaw('ADD_PARITY_1')));
+      };
+
+      const onSignature = (signature: string) => {
+        this.finalizeQrTransaction();
+
+        const sigBuf = Buffer.from(signature.substr(2), 'hex');
+
+        // Mimicking the way tx.sign() works
+        let v = sigBuf[64] + 27;
+
+        if (tx._chainId > 0) {
+          v += tx._chainId * 2 + 8;
+        }
+
+        tx.r = sigBuf.slice(0, 32);
+        tx.s = sigBuf.slice(32, 64);
+        tx.v = Buffer.from([v]);
+
+        resolve(tx.serialize());
+      };
+
+      this.setWalletQrTransaction(tx, from, onSignature, onCancel);
```

Comment:
Is there a reason this is a callback argument in constructor, instead of just resolving the promise with these variables? Would be nice if the wallet was more agnostic to where it gets used, so that you could use the same wallet instance to sign transactions with two different implementations.

Reply:
Yeah, this is the part I'm least happy about, and it goes along with how `QrSignerModal` is done. The callbacks go up to the container and are in turn placed in redux state for the modal to use.

The component that instantiates the wallet (`ParitySignerDecrypt` container in this case) has no knowledge of the modal, and is unmounted after unlocking so it cannot host the `QrSignerModal` itself, hence I'm passing these things over redux, which works but also feels over-engineered for my tastes.

Ideally I'd be able to pass the callbacks from the wallet to the modal as props without going through all these gymnastics, but I haven't been able to figure out how to make it do it any cleaner so far (my understanding of the existing code base is still pretty shallow).

Reply:
We came up with a migration strategy to move these things to Redux via slack DMs. Those changes probably won't land for a little while though.